.*: remove interleaved intents code

bananabrick opened the pull request:
In 21.2, we migrate interleaved intents to separated intents.
This pr removes most of the interleaved intent related code paths.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72536)
<!-- Reviewable:end -->

bananabrick committed:
.*: remove interleaved intents code

In 21.2, we migrate interleaved intents to separated intents.
This pr removes most of the interleaved intent related code paths.

Release note: None

bananabrick reviewed (commented):
This is good for review. I'm still failing one test related to intent interleaving iter, but it just looks like I need to understand some terminology to fix it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-:-MoCLTeD8e1VwscC8Fy7:b-wtaypt)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

bananabrick reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/intent_interleaving_iter_test.go, line 111 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoG868U0m_n-UztMJl_:-MoG868U0m_n-UztMJla:bdb7a2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/32d82ec6a54f9cbfa847e4b007a8c59aef1c23a0/pkg/storage/intent_interleaving_iter_test.go#L111)):*
> ```Go
> 	}
> 	rawMVCCKey := iter.UnsafeRawMVCCKey()
> 	if iter.IsCurIntentSeparated() {
> ```

The `doOps` function in this file doesn't use a `interleavedIntentIterator`, if the `interleave` flag passed into the function is `false`. It uses `iter1 = eng.NewMVCCIterator(MVCCKeyIterKind, opts)` instead. When `iter1` is used, `iter1.IsCurIntent` is always `false`, so the output between the interleaved case, and the non interleaved case, ends up being different for the `TestRandomizedIntentInterleavingIter`.
___
*[pkg/storage/intent_interleaving_iter_test.go, line 153 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoG8pD7ENOr-VDbi533:-MoG8pD7ENOr-VDbi534:b2mjg4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/32d82ec6a54f9cbfa847e4b007a8c59aef1c23a0/pkg/storage/intent_interleaving_iter_test.go#L153)):*
> ```Go
> 		return
> 	}
> 	if k1.Timestamp.IsEmpty() {
> ```

This is always false, when a non interleaving intent iterator is used in the `doOps` function, which results in a different output generated and the test failing.


<!-- Sent from Reviewable.io -->

bananabrick reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/intent_interleaving_iter_test.go, line 275 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoGGbuz7JcR-p6_3Xxy:-MoGGbv-84NgGhJilBG6:beny6o6) ([raw file](https://github.com/cockroachdb/cockroach/blob/32d82ec6a54f9cbfa847e4b007a8c59aef1c23a0/pkg/storage/intent_interleaving_iter_test.go#L275)):*
> ```Go
> 							// This is an abuse of PutUnversioned, but we know the
> 							// implementation and can be sure that it will not change the key.
> 							if err := batch.PutUnversioned(key, val); err != nil {
> ```

Note that I didn't modify the `testdata` files for this test. The files still have `meta` with a `ts` in the `mvcc` section, which leads to this `batch.PutUnversioned(key, val)` call here. The output of the test remains unchanged.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
Thanks for doing this!
Some minor comments.

> Reviewed 4 of 22 files at r1, 3 of 12 files at r2, 1 of 2 files at r3, 3 of 14 files at r9, 15 of 15 files at r10, 1 of 2 files at r12, 4 of 7 files at r13, 2 of 4 files at r14, 1 of 1 files at r15, 6 of 6 files at r16, 2 of 3 files at r17, 5 of 6 files at r19, 1 of 1 files at r21.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-:-MoeQXO8-4Mpc2q4cTzt:b-dcb7qc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick and @sumeerbhola)

___
*[pkg/storage/engine.go, line 157 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeDLkZBSIBQsq2dtAX:-MoeDLkZBSIBQsq2dtAY:bnv0i0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/engine.go#L157)):*
> ```Go
> 	ValueProto(msg protoutil.Message) error
> 	// Returns true iff Key() is positioned on an intent.
> 	IsCurIntent() bool
> ```

we shouldn't need this method any more.
I think the only remaining call is in intent_interleaving_iter_test.go, where the code block
```
	if iter.IsCurIntentSeparated() {
		...
```
can be changed to
```
   if engineKey.IsLockTableKey() {
     ....
```

___
*[pkg/storage/intent_interleaving_iter_test.go, line 111 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoG868U0m_n-UztMJl_:-MoePEtH7PJI5tKBB8-V:b-i8p94d) ([raw file](https://github.com/cockroachdb/cockroach/blob/32d82ec6a54f9cbfa847e4b007a8c59aef1c23a0/pkg/storage/intent_interleaving_iter_test.go#L111)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

The `doOps` function in this file doesn't use a `interleavedIntentIterator`, if the `interleave` flag passed into the function is `false`. It uses `iter1 = eng.NewMVCCIterator(MVCCKeyIterKind, opts)` instead. When `iter1` is used, `iter1.IsCurIntent` is always `false`, so the output between the interleaved case, and the non interleaved case, ends up being different for the `TestRandomizedIntentInterleavingIter`.
</blockquote></details>

The `interleave` parameter is somewhat confusing.
There are two kinds of interleave:
- physically interleaved intents: That is, does the engine store them interleaved. The interleave parameter in `writeRandomData` represents this physical interleaving. When set to true, it doesn't write any separated intents. When false, it does whatever `lockKV` and `mvccKV` is telling it to do, i.e., it will write both separated and interleaved intents.
- logically interleaved intents: This is what `intentInterleavingIter` does, and needs to be employed when physical interleaving is false. The interleave parameter in `doOps` represents this logical interleaving. Which is why `writeRandomData(..., false)` is paired with `doOps(..., true, ...)`.
___
*[pkg/storage/intent_interleaving_iter_test.go, line 275 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoGGbuz7JcR-p6_3Xxy:-MoeNZvNAeK-eW7oSS1c:b-91pjbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/32d82ec6a54f9cbfa847e4b007a8c59aef1c23a0/pkg/storage/intent_interleaving_iter_test.go#L275)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Note that I didn't modify the `testdata` files for this test. The files still have `meta` with a `ts` in the `mvcc` section, which leads to this `batch.PutUnversioned(key, val)` call here. The output of the test remains unchanged.
</blockquote></details>

We should add a comment here that since `intentInterleavingIter` still allows physically interleaved intents, we allow them in the test. A later PR should simplify intentInterleavingIter and this test.

For the same reason, I think you can keep `TestRandomizedIntentInterleavingIter` unchanged for now.

Doing this in a later PR works only because even though you have ripped out support for writing physically interleaved intents from the `MVCC*` functions, the tests in this file manually construct interleaved intents by directly writing to Pebble.
___
*[pkg/storage/intent_reader_writer.go, line 73 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeGyLF0DHICpOEUKah:-MoeGyLGDzQGKMIZ_gv7:bqnioko) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/intent_reader_writer.go#L73)):*
> ```Go
> 	key roachpb.Key,
> 	value []byte,
> 	state PrecedingIntentState,
> ```

we no longer need the state parameter.
___
*[pkg/storage/intent_reader_writer_test.go, line 236 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeHKRG2YvnsW3sQpv6:-MoeHKRG2YvnsW3sQpv7:b-tzmu58) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/intent_reader_writer_test.go#L236)):*
> ```Go
> 					return err.Error()
> 				}
> 				state := readPrecedingIntentState(t, d)
> ```

we won't need `readPrecedingIntentState` in this `PutIntent` case.
___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-MoeJ3A_-teTZBYIur1t:b-me7j9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
> ```Go
> //   false until there can never be v21.1 nodes.
> type txnDidNotUpdateMetaHelper struct {
> 	txnDidNotUpdateMeta bool
> ```

we should no longer need this `txnDidNotUpdateMeta` field in this helper.
___
*[pkg/storage/mvcc_history_test.go, line 95 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeK5rt2r68KFHWh_gg:-MoeK5rt2r68KFHWh_gh:b-p2r0kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc_history_test.go#L95)):*
> ```Go
> 	datadriven.Walk(t, "testdata/mvcc_histories", func(t *testing.T, path string) {
> 		// Default to random behavior wrt cluster version and separated
> 		// intents.
> ```

this comment can be deleted.
___
*[pkg/storage/pebble_iterator.go, line 571 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeKYdJ5mbNKcXx5sMq:-MoeKYdJ5mbNKcXx5sMr:b-k6ww98) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/pebble_iterator.go#L571)):*
> ```Go
> // IsCurIntent implements the MVCCIterator interface.
> func (p *pebbleIterator) IsCurIntent() bool {
> 	// todo(bananabrick): Can this be true?
> ```

this method should go away, as mentioned in a previous comment.


<!-- Sent from Reviewable.io -->

bananabrick reviewed (commented):
Thanks for taking a look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-:-MonubrV7rPsy-5mPV8a:bubxjul)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/storage/engine.go, line 157 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeDLkZBSIBQsq2dtAX:-MonwYaBA1Q-Xzyu1Tb3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/engine.go#L157)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

we shouldn't need this method any more.
I think the only remaining call is in intent_interleaving_iter_test.go, where the code block
```
	if iter.IsCurIntentSeparated() {
		...
```
can be changed to
```
   if engineKey.IsLockTableKey() {
     ....
```

</blockquote></details>

Done.
___
*[pkg/storage/intent_interleaving_iter_test.go, line 111 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoG868U0m_n-UztMJl_:-Moo47XD0C5BayYpN0oG:b-ucfkn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/32d82ec6a54f9cbfa847e4b007a8c59aef1c23a0/pkg/storage/intent_interleaving_iter_test.go#L111)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The `interleave` parameter is somewhat confusing.
There are two kinds of interleave:
- physically interleaved intents: That is, does the engine store them interleaved. The interleave parameter in `writeRandomData` represents this physical interleaving. When set to true, it doesn't write any separated intents. When false, it does whatever `lockKV` and `mvccKV` is telling it to do, i.e., it will write both separated and interleaved intents.
- logically interleaved intents: This is what `intentInterleavingIter` does, and needs to be employed when physical interleaving is false. The interleave parameter in `doOps` represents this logical interleaving. Which is why `writeRandomData(..., false)` is paired with `doOps(..., true, ...)`.

</blockquote></details>

Thanks.
___
*[pkg/storage/intent_reader_writer.go, line 73 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeGyLF0DHICpOEUKah:-Moo7YU88Lj8tf-Ijz0b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/intent_reader_writer.go#L73)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

we no longer need the state parameter.
</blockquote></details>

Done.
___
*[pkg/storage/intent_reader_writer_test.go, line 236 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeHKRG2YvnsW3sQpv6:-Moo7ZaICXFv90QXDqnz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/intent_reader_writer_test.go#L236)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

we won't need `readPrecedingIntentState` in this `PutIntent` case.
</blockquote></details>

Done.
___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-Moo8odCCsCX1SBu5dSO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

we should no longer need this `txnDidNotUpdateMeta` field in this helper.
</blockquote></details>

Done.
___
*[pkg/storage/mvcc_history_test.go, line 95 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeK5rt2r68KFHWh_gg:-Moo8wpo0ba15NKJje1H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc_history_test.go#L95)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this comment can be deleted.
</blockquote></details>

Done.
___
*[pkg/storage/pebble_iterator.go, line 571 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeKYdJ5mbNKcXx5sMq:-Moo8y_5BP210MNLrqGS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/pebble_iterator.go#L571)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this method should go away, as mentioned in a previous comment.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 2 of 22 files at r1, 2 of 12 files at r2, 1 of 6 files at r24, 2 of 2 files at r25, 6 of 10 files at r26, 3 of 3 files at r27, 5 of 5 files at r28, 2 of 2 files at r29, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-:-MosYvv87Yos_CB1D3H8:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick and @sumeerbhola)

___
*[pkg/storage/intent_interleaving_iter_test.go, line 521 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MosXgv6BmRuOIShTOxE:-MosXgv6BmRuOIShTOxF:b-dwt7v7) ([raw file](https://github.com/cockroachdb/cockroach/blob/32d82ec6a54f9cbfa847e4b007a8c59aef1c23a0/pkg/storage/intent_interleaving_iter_test.go#L521)):*
> ```Go
> 			val, err := protoutil.Marshal(&meta)
> 			require.NoError(t, err)
> 			isSeparated := rng.Int31n(2) == 0
> ```

Can you revert this change. For the same reason that we are testing physically interleaved intents in the rest of this file.
And leave a TODO somewhere near the top of intent_interleaving_iter.go stating that we should adjust the code to assume no interleaved intents, and return an error if one is found. When we fix that TODO we'll have to change these tests too.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
migration changes LGTM

nvb reviewed (commented):
> Reviewed 6 of 22 files at r1, 5 of 12 files at r2, 1 of 2 files at r3, 2 of 14 files at r9, 10 of 15 files at r10, 1 of 7 files at r13, 1 of 4 files at r14, 1 of 1 files at r15, 5 of 6 files at r16, 1 of 3 files at r17, 1 of 6 files at r19, 1 of 6 files at r24, 6 of 10 files at r26, 3 of 3 files at r27, 5 of 5 files at r28, 2 of 2 files at r29, 1 of 1 files at r30, 1 of 1 files at r31, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick and @sumeerbhola)

___
*[pkg/kv/kvserver/replica_rangefeed.go, line 395 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotU-615wTUMzNQ3hwn:-MotU-615wTUMzNQ3hwo:b6i8m5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/kv/kvserver/replica_rangefeed.go#L395)):*
> ```Go
> 		})
> 		return rangefeed.NewSeparatedIntentScanner(iter)
> ```

nit: stray line
___
*[pkg/server/sticky_engine.go, line 139 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotV7mz8_KbuJjpSoAb:-MotV7mz8_KbuJjpSoAc:b-i0vt2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/server/sticky_engine.go#L139)):*
> ```Go
> 		storage.EncryptionAtRest(spec.EncryptionOptions),
> 	}
> 	options = append(options, storage.ForStickyEngineTesting)
> ```

This can be moved into the declaration of the `options` slice.
___
*[pkg/storage/bench_test.go, line 285 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotVEl00zvdxEpsnusH:-MotVEl00zvdxEpsnusI:bdeuk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/bench_test.go#L285)):*
> ```Go
> 
> 	for _, sep := range []bool{true} {
> 		b.Run(fmt.Sprintf("separated=%t", sep), func(b *testing.B) {
> ```

These subtests throughout this file can go away.
___
*[pkg/storage/engine_test.go, line 1716 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotV_8-D5N94EoEI_GS:-MotV_8-D5N94EoEI_GT:bjub1b7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/engine_test.go#L1716)):*
> ```Go
> 
> 	for name, enableSeparatedIntents := range map[string]bool{"separated": true} {
> 		t.Run(name, func(t *testing.T) {
> ```

This subtest can go away.
___
*[pkg/storage/in_mem.go, line 59 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotVfbp7YAi0XqKH1u7:-MotVfbp7YAi0XqKH1u8:br83f04) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/in_mem.go#L59)):*
> ```Go
> // NewInMemForTesting is just like NewDefaultInMemForTesting but allows to
> // deterministically define whether it separates intents from MVCC data.
> func NewInMemForTesting(opts ...ConfigOption) Engine {
> ```

Is this now identical to `NewDefaultInMemForTesting`?
___
*[pkg/storage/intent_reader_writer.go, line 44 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotW9at2gzgIC3bsivL:-MotW9at2gzgIC3bsivM:b-2dw0dp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/intent_reader_writer.go#L44)):*
> ```Go
> func (idw intentDemuxWriter) ClearIntent(
> 	key roachpb.Key,
> 	state PrecedingIntentState,
> ```

Do we need this state parameter either? Seems like there's only one valid option.
___
*[pkg/storage/intent_reader_writer.go, line 72 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotWGwS-dK_whyELGja:-MotWGwS-dK_whyELGjb:b-jvzf10) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/intent_reader_writer.go#L72)):*
> ```Go
> 	ctx context.Context, key roachpb.Key, value []byte, txnUUID uuid.UUID, buf []byte,
> ) (_ []byte, _ error) {
> 	// The intent is either staying separated, or there
> ```

Does this comment provide any value now?
___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-MotWxgO0yfakRf7M-2Z:bf9b72y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Do we need the `PrecedingIntentState` field either? If not, can this entire helper struct go away?
___
*[pkg/storage/mvcc_stats_test.go, line 193 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotX6Sk4cs3kudZ6QCI:-MotX6Sk4cs3kudZ6QCJ:brg9173) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/mvcc_stats_test.go#L193)):*
> ```Go
> 				Txn:       &txn.TxnMeta,
> 			}).Size())
> 			var separatedIntentCount int64
> ```

Throughout this file, we no longer need the `separatedIntentCount` variable.
___
*[pkg/storage/pebble.go, line 922 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotXLnpCug-OJvhPvsp:-MotXLnpCug-OJvhPvsq:b-obiksr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/pebble.go#L922)):*
> ```Go
> 	key roachpb.Key, state PrecedingIntentState, txnDidNotUpdateMeta bool, txnUUID uuid.UUID,
> ) error {
> 	_, err :=
> ```

nit: the indentation on these calls looks pretty off now. Consider collapsing to a single line.
___
*[pkg/storage/pebble_mvcc_scanner_test.go, line 70 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotXWgYCZvMzBpqkd7n:-MotXWgYCZvMzBpqkd7o:b-vxspft) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/pebble_mvcc_scanner_test.go#L70)):*
> ```Go
> 
> 	for _, k := range keys {
> 		err = eng.PutIntent(
> ```

nit: same point about the indentation.


<!-- Sent from Reviewable.io -->

bananabrick reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/storage/intent_interleaving_iter_test.go, line 521 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MosXgv6BmRuOIShTOxE:-MotnNGh75f01G8FA7KM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32d82ec6a54f9cbfa847e4b007a8c59aef1c23a0/pkg/storage/intent_interleaving_iter_test.go#L521)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you revert this change. For the same reason that we are testing physically interleaved intents in the rest of this file.
And leave a TODO somewhere near the top of intent_interleaving_iter.go stating that we should adjust the code to assume no interleaved intents, and return an error if one is found. When we fix that TODO we'll have to change these tests too.

</blockquote></details>

Done.
___
*[pkg/storage/intent_reader_writer.go, line 44 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotW9at2gzgIC3bsivL:-MotnOFt--73WCTLJ5eu:b-g3eccz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/intent_reader_writer.go#L44)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need this state parameter either? Seems like there's only one valid option.
</blockquote></details>

I think we need it. We still have a state which signifies that there was no previous intent, in which case, `ClearIntent` should error.
___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-MotncNc4DiG8XNZQ76z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need the `PrecedingIntentState` field either? If not, can this entire helper struct go away?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

bananabrick reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-MotuEo94kfj1NqY8hfw:b-qk7ys6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Removing this changed the size of the meta field, which is causing test failures. Is it okay to have a size change here?


<!-- Sent from Reviewable.io -->

bananabrick reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-Motum8q1SiuoeM58cpA:b-en0ixw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Removing this changed the size of the meta field, which is causing test failures. Is it okay to have a size change here?
</blockquote></details>

This caused over 50+ test failures, which will have to be individually updated. I'm going to keep this around for now.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 14 of 15 files at r32, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/storage/intent_reader_writer.go, line 44 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotW9at2gzgIC3bsivL:-MotuoHJ8TApPdfNaN9U:b9vfku7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/intent_reader_writer.go#L44)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

I think we need it. We still have a state which signifies that there was no previous intent, in which case, `ClearIntent` should error.
</blockquote></details>

Does this need to be pushed through the `Writer` interface though? If a caller wants to make an assertion that they aren't calling `ClearIntent` on a non-existent intent, they still can.

Given that you've already removed `PrecedingIntentState` from the `PutIntent` method, I think this is a valuable thread to pull on. It may unravel all the way up to being able to delete the `PrecedingIntentState` type entirely. 
___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-Motviuc59OFbi6eDotO:bjtwalz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

This caused over 50+ test failures, which will have to be individually updated. I'm going to keep this around for now.
</blockquote></details>

Is this because we are no longer setting `meta.TxnDidNotUpdateMeta = &trueValue`?


<!-- Sent from Reviewable.io -->

bananabrick reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/storage/intent_reader_writer.go, line 44 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotW9at2gzgIC3bsivL:-MotzYe59_VsRubkprs7:b-1ia0da) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/intent_reader_writer.go#L44)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this need to be pushed through the `Writer` interface though? If a caller wants to make an assertion that they aren't calling `ClearIntent` on a non-existent intent, they still can.

Given that you've already removed `PrecedingIntentState` from the `PutIntent` method, I think this is a valuable thread to pull on. It may unravel all the way up to being able to delete the `PrecedingIntentState` type entirely. 
</blockquote></details>

We're preserving the old behaviour right now. We used to error out when the state wasn't separated or interleaved. Now we're erroring out when the state isn't interleaved.

Maybe we should have a second look at the assumptions ClearIntent makes, but right now this is just doing what we've always been doing, and I don't think this pr should aim to remove the `NoExistingIntent` state.
___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-Mou-75h1VbmpT3hkXHT:b-sk6rjm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this because we are no longer setting `meta.TxnDidNotUpdateMeta = &trueValue`?
</blockquote></details>

Yes. I could remove this, and update the tests, but considering there were 50 failures, it might not be worth it.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/storage/intent_reader_writer.go, line 44 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotW9at2gzgIC3bsivL:-Mou3QpX5MAXh3j236n2:b-6o7dvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/intent_reader_writer.go#L44)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

We're preserving the old behaviour right now. We used to error out when the state wasn't separated or interleaved. Now we're erroring out when the state isn't interleaved.

Maybe we should have a second look at the assumptions ClearIntent makes, but right now this is just doing what we've always been doing, and I don't think this pr should aim to remove the `NoExistingIntent` state.
</blockquote></details>

This is another case where we should strive to eliminate complexity if it is no longer needed. Unwarranted complexity makes code harder to read and leads to bugs. It also places a burden on the implementor of interfaces if that complexity crosses interface boundaries like it does here. Every implementor of `Writer` still needs to handle all possible `PrecedingIntentState`s, even if none of the callers need them to.

If we take a look at the three uses of `ClearIntent`, which are all in `mvccResolveWriteIntent`, we see that none of them are reachable with a `NoExistingIntent` `PrecedingIntentState`. We now have `precedingIntentState := ExistingIntentSeparated` at the beginning of that function. So it does look possible to clean this up. 

But again, that doesn't need to be in this PR! I just want to make sure we don't forget this. If we're going to leave this around though, let's open an issue to track the additional cleanup.

Also, if we're going to leave that around for the time being, I'd recommend replacing `precedingIntentState := ExistingIntentSeparated` with `const precedingIntentState = ExistingIntentSeparated` to make it extra clear that the argument is a constant that can only take on one value.
___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-Mou31Ks-u8CXWmNXYyZ:bu6jbbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Yes. I could remove this, and update the tests, but considering there were 50 failures, it might not be worth it.
</blockquote></details>

Not worth it in this PR, or not worth it at all? There's a large amount of complexity here (just look at the size of that comment!), so it does seem valuable to remove it if possible. But that certainly doesn't need to take place in this PR, since it's already quite large. If we're going to hold off, let's open an issue.


<!-- Sent from Reviewable.io -->

bananabrick reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/storage/intent_reader_writer.go, line 44 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotW9at2gzgIC3bsivL:-MoubYOU9LlfbfnPgM0L:bc2qa0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/intent_reader_writer.go#L44)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is another case where we should strive to eliminate complexity if it is no longer needed. Unwarranted complexity makes code harder to read and leads to bugs. It also places a burden on the implementor of interfaces if that complexity crosses interface boundaries like it does here. Every implementor of `Writer` still needs to handle all possible `PrecedingIntentState`s, even if none of the callers need them to.

If we take a look at the three uses of `ClearIntent`, which are all in `mvccResolveWriteIntent`, we see that none of them are reachable with a `NoExistingIntent` `PrecedingIntentState`. We now have `precedingIntentState := ExistingIntentSeparated` at the beginning of that function. So it does look possible to clean this up. 

But again, that doesn't need to be in this PR! I just want to make sure we don't forget this. If we're going to leave this around though, let's open an issue to track the additional cleanup.

Also, if we're going to leave that around for the time being, I'd recommend replacing `precedingIntentState := ExistingIntentSeparated` with `const precedingIntentState = ExistingIntentSeparated` to make it extra clear that the argument is a constant that can only take on one value.

</blockquote></details>

You're right. I'll remove those in a follow up pr. I can do it in this one too, but it's already become difficult to get the build passing here.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-Mp73CxKAnnfNrP-a66O:bxztwgz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Not worth it in this PR, or not worth it at all? There's a large amount of complexity here (just look at the size of that comment!), so it does seem valuable to remove it if possible. But that certainly doesn't need to take place in this PR, since it's already quite large. If we're going to hold off, let's open an issue.
</blockquote></details>

We still need to populate MVCCMetatadata.TxnDidNotUpdateMeta, which is what PrecedingIntentState is for.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/storage/intent_reader_writer.go, line 44 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MotW9at2gzgIC3bsivL:-Mp7BPGA650J2wdIjNFp:b-94p0e4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d66cde787c836b044517678eee55a563eee9760/pkg/storage/intent_reader_writer.go#L44)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

You're right. I'll remove those in a follow up pr. I can do it in this one too, but it's already become difficult to get the build passing here.
</blockquote></details>

A follow-up PR is completely fine. Thanks.
___
*[pkg/storage/mvcc.go, line 1144 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-MoeJ3A_-teTZBYIur1s:-Mp799MN6o8PMacHcZPm:b-k07jcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2c6e978258667fb639c3bd195ed02b4af6470a5/pkg/storage/mvcc.go#L1144)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We still need to populate MVCCMetatadata.TxnDidNotUpdateMeta, which is what PrecedingIntentState is for.
</blockquote></details>

Do you think this `txnDidNotUpdateMetaHelper` is too heavy-weight now that it's only used once in `putIntentMeta`? It seems worth exploring whether we can inline this assignment to `MVCCMetatadata.TxnDidNotUpdateMeta` logic into the two callers without any loss of clarity.

The strongest argument for keeping it might actually be the associated comment, which is nice. But the comment is also mostly talking about the migration, which is no longer relevant and can be removed.

The other argument for keeping it is that it prevents a caller of `putIntentMeta` from forgetting to update `MVCCMetatadata.TxnDidNotUpdateMeta`, but the same could be accomplished with a `bool` parameter.


<!-- Sent from Reviewable.io -->

sumeerbhola commented:
> Do you think this txnDidNotUpdateMetaHelper is too heavy-weight now that it's only used once in putIntentMeta?

It probably is.
@bananabrick could you add a TODO(sumeer) there to potentially simplify the code.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 3 files at r6, 1 of 1 files at r34, 1 of 1 files at r35, 1 of 1 files at r36, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72536#-:-MpECVuYEp58gPTQjLgn:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @bananabrick and @sumeerbhola)



<!-- Sent from Reviewable.io -->

bananabrick commented:
bors r+
